// victory-core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Box: true,
  addEvents: true,
  Collection: true,
  Data: true,
  DefaultTransitions: true,
  Domain: true,
  Events: true,
  Helpers: true,
  Immutable: true,
  LabelHelpers: true,
  Log: true,
  PropTypes: true,
  Scale: true,
  Selection: true,
  Style: true,
  TextSize: true,
  Transitions: true,
  UserProps: true,
  CommonProps: true,
  Wrapper: true,
  Axis: true,
  Hooks: true,
  LineHelpers: true,
  Timer: true,
  PointPathHelpers: true,
  TimerContext: true
};
exports.Axis = void 0;
Object.defineProperty(exports, "Box", {
  enumerable: true,
  get: function () {
    return _victoryPrimitives.Border;
  }
});
exports.TextSize = exports.Style = exports.Selection = exports.Scale = exports.PropTypes = exports.PointPathHelpers = exports.Log = exports.LineHelpers = exports.LabelHelpers = exports.Immutable = exports.Hooks = exports.Helpers = exports.Events = exports.Domain = exports.DefaultTransitions = exports.Data = exports.CommonProps = exports.Collection = void 0;
Object.defineProperty(exports, "Timer", {
  enumerable: true,
  get: function () {
    return _timer.default;
  }
});
Object.defineProperty(exports, "TimerContext", {
  enumerable: true,
  get: function () {
    return _timerContext.default;
  }
});
exports.Wrapper = exports.UserProps = exports.Transitions = void 0;
Object.defineProperty(exports, "addEvents", {
  enumerable: true,
  get: function () {
    return _addEvents.default;
  }
});

var _victoryAccessibleGroup = require("./victory-accessible-group/victory-accessible-group");

Object.keys(_victoryAccessibleGroup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryAccessibleGroup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryAccessibleGroup[key];
    }
  });
});

var _victoryAnimation = require("./victory-animation/victory-animation");

Object.keys(_victoryAnimation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryAnimation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryAnimation[key];
    }
  });
});

var _victoryContainer = require("./victory-container/victory-container");

Object.keys(_victoryContainer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryContainer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryContainer[key];
    }
  });
});

var _victoryLabel = require("./victory-label/victory-label");

Object.keys(_victoryLabel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryLabel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryLabel[key];
    }
  });
});

var _victoryTransition = require("./victory-transition/victory-transition");

Object.keys(_victoryTransition).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryTransition[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryTransition[key];
    }
  });
});

var _victoryClipContainer = require("./victory-clip-container/victory-clip-container");

Object.keys(_victoryClipContainer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryClipContainer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryClipContainer[key];
    }
  });
});

var _victoryTheme = require("./victory-theme/victory-theme");

Object.keys(_victoryTheme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryTheme[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryTheme[key];
    }
  });
});

var _portal = require("./victory-portal/portal");

Object.keys(_portal).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _portal[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _portal[key];
    }
  });
});

var _portalContext = require("./victory-portal/portal-context");

Object.keys(_portalContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _portalContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _portalContext[key];
    }
  });
});

var _victoryPortal = require("./victory-portal/victory-portal");

Object.keys(_victoryPortal).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryPortal[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryPortal[key];
    }
  });
});

var _victoryPrimitives = require("./victory-primitives");

Object.keys(_victoryPrimitives).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _victoryPrimitives[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _victoryPrimitives[key];
    }
  });
});

var _addEvents = _interopRequireDefault(require("./victory-util/add-events"));

var _Collection = _interopRequireWildcard(require("./victory-util/collection"));

exports.Collection = _Collection;

var _Data = _interopRequireWildcard(require("./victory-util/data"));

exports.Data = _Data;

var _DefaultTransitions = _interopRequireWildcard(require("./victory-util/default-transitions"));

exports.DefaultTransitions = _DefaultTransitions;

var _Domain = _interopRequireWildcard(require("./victory-util/domain"));

exports.Domain = _Domain;

var _Events = _interopRequireWildcard(require("./victory-util/events"));

exports.Events = _Events;

var _Helpers = _interopRequireWildcard(require("./victory-util/helpers"));

exports.Helpers = _Helpers;

var _Immutable = _interopRequireWildcard(require("./victory-util/immutable"));

exports.Immutable = _Immutable;

var _LabelHelpers = _interopRequireWildcard(require("./victory-util/label-helpers"));

exports.LabelHelpers = _LabelHelpers;

var _Log = _interopRequireWildcard(require("./victory-util/log"));

exports.Log = _Log;

var _PropTypes = _interopRequireWildcard(require("./victory-util/prop-types"));

exports.PropTypes = _PropTypes;

var _Scale = _interopRequireWildcard(require("./victory-util/scale"));

exports.Scale = _Scale;

var _Selection = _interopRequireWildcard(require("./victory-util/selection"));

exports.Selection = _Selection;

var _Style = _interopRequireWildcard(require("./victory-util/style"));

exports.Style = _Style;

var _TextSize = _interopRequireWildcard(require("./victory-util/textsize"));

exports.TextSize = _TextSize;

var _Transitions = _interopRequireWildcard(require("./victory-util/transitions"));

exports.Transitions = _Transitions;

var _UserProps = _interopRequireWildcard(require("./victory-util/user-props"));

exports.UserProps = _UserProps;

var _CommonProps = _interopRequireWildcard(require("./victory-util/common-props"));

exports.CommonProps = _CommonProps;

var _Wrapper = _interopRequireWildcard(require("./victory-util/wrapper"));

exports.Wrapper = _Wrapper;

var _Axis = _interopRequireWildcard(require("./victory-util/axis"));

exports.Axis = _Axis;

var _Hooks = _interopRequireWildcard(require("./victory-util/hooks"));

exports.Hooks = _Hooks;

var _LineHelpers = _interopRequireWildcard(require("./victory-util/line-helpers"));

exports.LineHelpers = _LineHelpers;

var _timer = _interopRequireDefault(require("./victory-util/timer"));

var _PointPathHelpers = _interopRequireWildcard(require("./victory-util/point-path-helpers"));

exports.PointPathHelpers = _PointPathHelpers;

var _timerContext = _interopRequireDefault(require("./victory-util/timer-context"));

var _types = require("./victory-util/types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
